pmlc/dialect/stripe/canon.cc

```diff
@@ -0,0 +1,164 @@
+
+  mlir::PatternMatchResult match(AffineMulOp op) const final {
+    auto val = mlir::dyn_cast<AffineConstOp>(op.input()->getDefiningOp());
+    if (!val) {
+      return matchFailure();
+    }
+    return matchSuccess();
+  }
+
+  void rewrite(AffineMulOp op, mlir::PatternRewriter& rewriter) const final {  // NOLINT(runtime/references)
+    auto val = mlir::dyn_cast<AffineConstOp>(op.input()->getDefiningOp());
+    auto constValue = val.value().getSExtValue();
+    if (constValue == 0) {
+      rewriter.replaceOp(op, op.input());  // A handy source of a zero
+    } else {
+      auto maybeDead = llvm::SmallVector<mlir::Value*, 1>{op.input()};
+      rewriter.replaceOpWithNewOp<AffineConstOp>(maybeDead, op, rewriter.getType<AffineType>(),
+                                                 rewriter.getI64IntegerAttr((op.scale() * constValue).getSExtValue()));
+    }
+  }
+};
+
+struct SimplifyMulConstZero final : public mlir::OpRewritePattern<AffineMulOp> {
+  explicit SimplifyMulConstZero(mlir::MLIRContext* context) : OpRewritePattern<AffineMulOp>(context) {}
+
+  mlir::PatternMatchResult match(AffineMulOp op) const final {
+    if (!op.scale()) {
+      return matchSuccess();
+    }
+    return matchFailure();
+  }
+
+  void rewrite(AffineMulOp op, mlir::PatternRewriter& rewriter) const final {  // NOLINT(runtime/references)
+    auto maybeDead = llvm::SmallVector<mlir::Value*, 1>{op.input()};
+    rewriter.replaceOpWithNewOp<AffineConstOp>(maybeDead, op, rewriter.getType<AffineType>(),
+                                               rewriter.getI64IntegerAttr(0));
+  }
+};
+
+struct SimplifyNopRefines final : public mlir::OpRewritePattern<RefineOp> {
+  explicit SimplifyNopRefines(mlir::MLIRContext* context) : OpRewritePattern<RefineOp>(context) {}
+
+  mlir::PatternMatchResult match(RefineOp op) const final {
+    for (auto* offset : op.offsets()) {
+      auto val = mlir::dyn_cast_or_null<AffineConstOp>(offset->getDefiningOp());
+      if (!val || !!val.value()) {
+        return matchFailure();
+      }
+    }
+    return matchSuccess();
+  }
+
+  void rewrite(RefineOp op, mlir::PatternRewriter& rewriter) const final {  // NOLINT(runtime/references)
+    rewriter.replaceOp(op, op.in());
+  }
+};
+
+}  // namespace
+
+void AffineAddOp::getCanonicalizationPatterns(OwningRewritePatternList& results, MLIRContext* context) {
+  results.insert<SimplifyAddNothing>(context);
+  results.insert<SimplifyAddSingle>(context);
+  results.insert<SimplifyAddZero>(context);
+}
+
+void AffineMulOp::getCanonicalizationPatterns(OwningRewritePatternList& results, MLIRContext* context) {
+  results.insert<SimplifyMulConstOne>(context);
+  results.insert<SimplifyMulConstValue>(context);
+  results.insert<SimplifyMulConstZero>(context);
+}
+
+void RefineOp::getCanonicalizationPatterns(OwningRewritePatternList& results, MLIRContext* context) {
+  results.insert<SimplifyNopRefines>(context);
+}
```

Comment:
Actually, I think all of these are really folders instead of canonicalizers (folding happens inside the canonicalize pass tho).

Reply:
See:
https://github.com/plaidml/plaidml/blob/master/pmlc/dialect/eltwise/ops.td#L220
https://github.com/plaidml/plaidml/blob/master/pmlc/dialect/eltwise/ops.cc#L237

Reply:
Most of them are, definitely (although not refines, I think?).  I'd missed `hasFolder` in the mlir dialects -- rewriting now... :-)

Reply:
And then for tests:
https://github.com/plaidml/plaidml/blob/master/pmlc/dialect/eltwise/tests/constant-fold.mlir